src/common/Utilities/TaskScheduler.h

```diff
@@ -0,0 +1,639 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-GPL2
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
+ */
+
+#ifndef _TASK_SCHEDULER_H_
+#define _TASK_SCHEDULER_H_
+
+#include <optional>
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <vector>
+#include <queue>
+#include <memory>
+#include <optional>
+#include <set>
+#include <utility>
+
+#include "Util.h"
+
+class TaskContext;
+
+/// The TaskScheduler class provides the ability to schedule std::function's in the near future.
+/// Use TaskScheduler::Update to update the scheduler.
+/// Popular methods are:
+/// * Schedule (Schedules a std::function which will be executed in the near future).
+/// * Schedules an asynchronous function which will be executed at the next update tick.
+/// * Cancel, Delay & Reschedule (Methods to manipulate already scheduled tasks).
+/// Tasks are organized in groups (uint), multiple tasks can have the same group id,
+/// you can provide a group or not, but keep in mind that you can only manipulate specific tasks through its group id!
+/// Tasks callbacks use the function signature void(TaskContext) where TaskContext provides
+/// access to the function schedule plan which makes it possible to repeat the task
+/// with the same duration or a new one.
+/// It also provides access to the repeat counter which is useful for task that repeat itself often
+/// but behave different every time (spoken event dialogs for example).
+class TaskScheduler
+{
+    friend class TaskContext;
+
+    // Time definitions (use steady clock)
+    typedef std::chrono::steady_clock clock_t;
+    typedef clock_t::time_point timepoint_t;
+    typedef clock_t::duration duration_t;
+
+    // Task group type
+    typedef uint32 group_t;
+    // Task repeated type
+    typedef uint32 repeated_t;
+    // Task handle type
+    typedef std::function<void(TaskContext)> task_handler_t;
+    // Predicate type
+    typedef std::function<bool()> predicate_t;
+    // Success handle type
+    typedef std::function<void()> success_t;
+
+    class Task
+    {
+        friend class TaskContext;
+        friend class TaskScheduler;
+
+        timepoint_t _end;
+        duration_t _duration;
+        std::optional<group_t> _group;
```

Comment:
what is this optional ? not found. need Boost library ?

Reply:
boot no need for this

Reply:
https://en.cppreference.com/w/cpp/utility/optional

Reply:
update yr data base C++17 released